### ISSUE TITLE
Cannot disable redirects with `maxRedirects: 0`

### ISSUE THREAD

**Issue by Shayan-To** [axios/axios#3924]
#### Describe the bug
I'm using axios to send a login request to an old back-end that always returns a 302 no matter it being successful or not. I don't want to actually follow the redirect that server is suggesting (as the front-end is being rewritten from scratch and the location is non-existant). And also, I'm required to have the headers that the server is returning with the 302 response.

In short: I don't want the redirects to be followed and I want to process the original 302 response.

#### To Reproduce
This works as expected in nodejs, but it does not work in a browser. (I had to load axios in https://httpstat.us/ and run the code in the console to prevent CORS errors.)

```js
await axios.request({
    method: "GET",
    url: "https://httpstat.us/302",
    maxRedirects: 0,
    validateStatus: null,
});
```

The returned response in the browser is the redirected response with status code 200.

#### Expected behavior
No redirects are followed in the browser (like in nodejs, and as described in the documentation), and the response status code returned is 302.

#### Environment
 - Axios Version 0.21.1 (latest)
 - Adapter I don't know.
 - Browser Firefox
 - Browser Version 90.0.2 (latest)
 - Node.js Version N/A
 - OS: Windows 10 Build 19043.928
 - Additional Library Versions N/A

#### Additional context/Screenshots
---

**Comment by Shayan-To** [comment]
Just tested on Microsoft Edge version 92.0.902.62, the same results, the bug exists.

**Comment by Shayan-To** [comment]
@IYaksha Whatever the doc says, I don't think that it is normal that I cannot disable redirects. I have never used a library that I couldn't do that with. Also, `fetch` supports that.

**Comment by IYaksha** [comment]
@Shayan-To Yes you are right that ```fetch``` supports it. The main reason why ```maxRedirects``` doesn't work is that axios used ```xhr``` adapter for browser and ```http``` adapter for server. Since axios use ```xhr``` adapter, which is based on ```XMLHttpRequest``` and since ```XMLHttpRequest``` doesn't have the feature to disable redirects that's why it doesn't work in axios on browser. ```fetch``` api is new so it added this feature. If someone wrote a ```fetch``` based adapter in axios then only I think your current issue will get solved because this is a limitation of ```xhr``` itself.

**Comment by alpgul** [comment]
The max Redirects default value should be 0. I've been trying to pull Status Code 302 for 2 days. Today I saw that the default value is 5. It was always returning 200 code.

**Comment by 0x0a0d** [comment]
https://stackoverflow.com/a/343359/14749432
Looks like it's not axios bug

**Comment by antoninbeaufort** [comment]
As of January 18th 2023 maxRedirects: 0 works as expected

**Comment by andreynovikov** [comment]
No, it's not.

**Comment by antoninbeaufort** [comment]
> No, it's not.

Are you on the latest ? I'm not the maintener or anyone with interest I've just encountered the same problem of redirection and tried on latest with maxRedirects:0 and now I get the first 302 response as expected

**Comment by DigitalBrainJS** [comment]
Yes, maxRedirects is still only supported for the HTTP adapter (nodejs). XMLHttpRequest has no way to manage redirects.

**Comment by wanglin86769** [comment]
maxRedirects: 0 works for me on Node.js but validateStatus config is also needed to accept 302 status code.

**Comment by anetbnd** [comment]
That this does not work is really a big issue. If you have APIs which automatically redirect a request to a login page in case the user session has expired, you have no chance with axios to detect this error case and start an appropriate error handling.

**Comment by pauldraper** [comment]
This works for Node.js.

<details>
<summary>Example</summary>

```js
const axios = require("axios");

const url = "http://google.com/";
const config = { maxRedirects: 0, validateStatus: () => true };
axios.get(url, config).then(response => console.log(response.status));
```

```txt
301
```
</details>

It does not work for the browser.

This is because axios uses XMLHttpRequest, which always follows redirects. See #1219 for supporting fetch, which allows customizing the redirect behavior.

**Comment by anetbnd** [comment]
I know that XMLHttpRequest is the root cause for not supporting the usecase of catching login redirects by axios. This makes it even more unfortunate, since it seems to be unlikely that the request in your link (#1219 ) will ever be implemented 😢
